File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-8.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {  
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if(root == null){
            return false;
        }

        boolean isValid = false;

        if(root.val == subRoot.val){
            isValid = isSameTree(root, subRoot);
        }

        boolean leftResult = isSubtree(root.left, subRoot);
        boolean rightResult = isSubtree(root.right, subRoot);

        if(isValid || leftResult || rightResult){
            return true;
        }

        return false;
    }

    public boolean isSameTree(TreeNode p, TreeNode q){
        if(p == null && q == null){
            return true;
        }

        if((p != null && q == null) || (p == null && q != null) || (p.val != q.val)){
            return false;
        }

        boolean leftResult = isSameTree(p.left, q.left);
        boolean rightResult = isSameTree(p.right, q.right);

        if(leftResult == false || rightResult == false){
            return false;
        }

        return true;
    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-8.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.isValid(root, float('-inf'), float('inf'))
        
    def isValid(self, root, left, right):
        if root is None:
            return True
        
        is_valid = False

        if left < root.val and root.val < right:
            is_valid = True

        left_result = self.isValid(root.left, left, root.val)
        right_result = self.isValid(root.right, root.val, right)

        return left_result and is_valid and right_result